internal/controllers/module_nmc_reconciler.go

```diff
@@ -439,6 +450,80 @@ func (mnrh *moduleNMCReconcilerHelper) moduleUpdateWorkerPodsStatus(ctx context.
 	return mnrh.client.Status().Patch(ctx, mod, client.MergeFrom(unmodifiedMod))
 }
 
+type namespaceHelper interface {
+	setLabel(ctx context.Context, name string) error
+	tryRemovingLabel(ctx context.Context, name, moduleName string) error
+}
+
+type namespaceHelperImpl struct {
+	client client.Client
+}
+
+func newNamespaceHelper(client client.Client) namespaceHelper {
+	return &namespaceHelperImpl{client: client}
+}
+
+func (h *namespaceHelperImpl) setLabel(ctx context.Context, name string) error {
+	logger := log.FromContext(ctx)
+
+	ns := v1.Namespace{}
+
+	if err := h.client.Get(ctx, types.NamespacedName{Name: name}, &ns); err != nil {
+		return fmt.Errorf("could not get namespace %s: %v", name, err)
+	}
+
+	if !meta.HasLabel(&ns, constants.NamespaceLabelKey) {
+		nsCopy := ns.DeepCopy()
+
+		logger.Info("Setting namespace label")
+		meta.SetLabel(&ns, constants.NamespaceLabelKey, "")
+
+		return h.client.Patch(ctx, &ns, client.MergeFrom(nsCopy))
+	}
+
+	return nil
+}
+
+func (h *namespaceHelperImpl) tryRemovingLabel(ctx context.Context, name, moduleName string) error {
+	logger := log.FromContext(ctx)
+
+	modList := kmmv1beta1.ModuleList{}
+
+	opt := client.InNamespace(name)
+
+	if err := h.client.List(ctx, &modList, opt); err != nil {
+		return fmt.Errorf("could not list modules in namespace %s: %v", name, err)
+	}
+
+	if count := len(modList.Items); count > 1 {
+		logger.Info("Namespace still contains modules; not removing the label", "count", count)
```

Comment:
Aren't we going to spam the lobs until the user removes the module?

Reply:
since this log will be issued only during module deletion, and it is supposed to be short-time occurrence, i think it should be OK. This log will be very helpful in case of some weird behavior that we will need to debug